specs/vep-2448-vdk-run-logs-simplified-and-readable/README.md

```diff
@@ -366,10 +366,290 @@ or slowed down due to progress bar updates. This is really relevant for
+        - Each time a new data is added in queue we can use tracker.add (# payloads, likely in _send method)
+            - For this new hook could be added every time a new payload is being send (ingest_send_payload hook)
+        - Each time is sent, we update progress. (we could use existing  hook : [post_ingest_process](https://github.com/vmware/versatile-data-kit/blob/main/projects/vdk-core/src/vdk/api/plugin/plugin_input.py#L486) )
+
+The ProgressTracker would be exposed and available in JobInput interface so it can be used by DAGs .
+
+For DAG it can be used  in this way
+```python
+current_step_progress_tracker = job_input.get_progress_tracker(
+
+dag_tracker = current_step_progress_tracker.create_sub_tracker()
+```
+
+The interface expose to the user and to plugins would look like this
+```python
+class IProgressTracker:
+    """
+    A class for tracking the progress of a task and its sub-tasks.
+    """
+
+    def add(self, new_iterations: int):
+        """
+        Adds new iterations to the total count for this tracker.
+
+        :param new_iterations: The number of new iterations to add.
+        """
+
+
+    def update_progress(self, iterations: int = 1):
+        """
+        Update progress with how many iteration have we completed.
+        Generally only childless trackers should be directly updated.
+        Parent trackers are automatically updated by their children.
+        """
+
+    def create_sub_tracker(self, title: str, total_iterations: int) -> ProgressTracker:
+        """
+        Create a new child tracker. The child tracker will point to its parent.
+        So any iterations added to the child tracker are automatically tracked by the parent as well.
+        And will update the parent automatically with any progress.
+        Any tracked metric is propagated to the parent as well
+        """
+
+    def track_metric(self, key: str, value: float):
+        """
+        The tracked metric must be cumulative since it's aggregated in the parent.
+        Non-cumulative metrics (like rate, latency, mean, median) will not provide result.
+        Example of cumulative metrics are  records processed, errors, time spent, number of requests, successful requests, failed requests.
+        """
+        self._metrics[key] = self._metrics.get(key, 0) + value
+        if self._parent:
+            self._parent.track_metric(key, value)
+
+    def get_metrics(self) -> Dict[str, int]:
+        """
+        Return dictionary with currently collected metrics about what is being tracked.
+        """
+```
+
+###### ProgressStrategy
+An abstract class in vdk-core that specifies the contract for strategy implementations in different environments.
+The basic logging strategy will be implemented in vdk-core while other strategies like CLI and notebook will be implemented in separate plugins like vdk-tqdm and vdk-jupyter.
```

Comment:
I think we could use a more involved explanation of what a ProgressStrategy is; this paragraph does not make it clear to me.

Reply:
The ProgressStrategy class in vdk-core serves as an abstract contract that defines how progress tracking will be communicated across different environments. The objective is to offer a flexible yet standardized way of implementing progress displays/presenters. be it logging information to stdout, showing a CLI-based progress bar, or rendering progress in Jupyter notebooks.

Initial support for basic logging will be built (in vdk-core), while more specialized strategies for CLI or notebook environments can be provided via separate plugins like vdk-tqdm for terminal-based progress bars or vdk-jupyter for notebook-specific progress indicators , etc. 


I am wondering if naming it ProgressDisplay or ProgressPresenter would make it clearer.